Pod/Classes/PINURLSessionManager.m

```diff
@@ -64,19 +67,37 @@ - (void)unlock
 
 - (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data
 {
-    [self.delegate didReceiveData:data forTask:dataTask];
+    [self lock];
+    dispatch_queue_t delegateQueue = self.delegateQueues[@(dataTask.taskIdentifier)];
+    [self unlock];
+    
+    __weak typeof(self) weakSelf = self;
+    dispatch_async(delegateQueue, ^{
+        [weakSelf.delegate didReceiveData:data forTask:dataTask];
+    });
 }
 
 - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
 {
-    [self.delegate didCompleteTask:task withError:error];
     [self lock];
-        void (^completionHandler)(NSURLResponse *, NSError *) = self.completions[@(task.taskIdentifier)];
-        [self.completions removeObjectForKey:@(task.taskIdentifier)];
+    dispatch_queue_t delegateQueue = self.delegateQueues[@(task.taskIdentifier)];
     [self unlock];
-    if (completionHandler) {
-        completionHandler(task.response, error);
-    }
+    
+    __weak typeof(self) weakSelf = self;
+    dispatch_async(delegateQueue, ^{
```

Comment:
No, delegate queue is created with the task and only removed in the following dispatch. At least that's *supposed* to be how it works.